connexion/middleware/security.py

```diff
@@ -0,0 +1,229 @@
+import logging
+import pathlib
+import typing as t
+from collections import defaultdict
+
+from starlette.types import ASGIApp, Receive, Scope, Send
+
+from connexion.apis.abstract import AbstractSpecAPI
+from connexion.exceptions import MissingMiddleware
+from connexion.http_facts import METHODS
+from connexion.lifecycle import MiddlewareRequest
+from connexion.middleware import AppMiddleware
+from connexion.security import SecurityHandlerFactory
+
+logger = logging.getLogger("connexion.middleware.security")
+
+
+class SecurityMiddleware(AppMiddleware):
+    """Middleware to check if operation is accessible on scope."""
+
+    def __init__(self, app: ASGIApp) -> None:
+        self.app = app
+        self.apis: t.Dict[str, SecurityAPI] = {}
+
+    def add_api(self, specification: t.Union[pathlib.Path, str, dict], **kwargs) -> None:
+        api = SecurityAPI(specification, **kwargs)
+        self.apis[api.base_path] = api
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send):
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        try:
+            connexion_context = scope['extensions']['connexion_routing']
+        except KeyError:
+            raise MissingMiddleware('Could not find operation_id in scope. Please make sure '
+                                    'you have a routing middleware registered upstream. ')
+
+        api_base_path = connexion_context.get('api_base_path')
+        if api_base_path:
+            api = self.apis[api_base_path]
+            operation_id = connexion_context.get('operation_id')
+            try:
+                operation = api.operations[operation_id]
+            except KeyError:
+                pass
```

Comment:
What is the reason why we silently skip the Security check when the operation_id was not found in the registered APIs? Would still put a warning/info log here, wdyt?

Reply:
The `operation_id` can be `None` for operations not specified in the spec. A default `SecurityOperation` is only registered when using `auth_all_paths`, so this can throw a `KeyError`.

Refactored this and included debug log and a check for unknown `operation_id`s.